### ISSUE TITLE
Unable to set cookies with batching operations

### ISSUE THREAD

**Issue by macs91** [hasura/graphql-engine#4015]
When enabling apollo client batching operations Hasura won't forward cookie header from operations in remote schema

**Comment by tirumaraiselvan** [comment]
@macs91 You are right. It was indeed unclear at the time what to do with remote schema response headers because of conflicts etc. You can see the PR description here: https://github.com/hasura/graphql-engine/pull/3490

Let's use this thread to discuss the issue. Pinging @paf31 @0x777 @lexi-lambda

**Comment by paf31** [comment]
Here's a brief summary of the points in that PR:

- We want to treat batched operations as if they are equivalent to the same set of operations run separately, and that is how we'll give meaning to batched queries.
- One question is, should we treat the separate operations as if they were run sequentially or in parallel? There can't be any dependencies between the operations (that is, operations cannot depend on the results of other operations) because otherwise they wouldn't end up in the same batch in the first place, so we may as well treat them as if they were issued in parallel. Therefore, if we see errors, the behavior should be to collect all of the errors together, not to stop execution on the first error.

In the case where we have several `Set-Cookie` headers coming back from the server in a batch, we need to decide how to merge them together into a single header. I think the best option (if we want to support cookies in batched operations at all) is as follows:

- Continue to execute the operations in sequence
- For each operation, send the same `Cookie` header we received from the client, i.e. don't update `Cookie` headers based on the `Set-Cookie` responses we see from the remote server
- For each response, parse the `Set-Cookie` response header, and combine those based on a "last wins per header key" approach.

i.e. we continue to act as if we're sending the requests in parallel, but in order to know the order that we receive the responses, we also continue to _actually_ send them in sequence, and combine `Set-Cookie` headers accordingly.
